src/spi/blocking.rs

```diff
@@ -1,98 +1,76 @@
 //! Blocking SPI API
 
-/// Blocking transfer with separate buffers
-pub trait Transfer<W = u8> {
-    /// Error type
-    type Error: crate::spi::Error;
-
-    /// Writes and reads simultaneously. `write` is written to the slave on MOSI and
-    /// words received on MISO are stored in `read`.
-    ///
-    /// It is allowed for `read` and `write` to have different lengths, even zero length.
-    /// The transfer runs for `max(read.len(), write.len())` words. If `read` is shorter,
-    /// incoming words after `read` has been filled will be discarded. If `write` is shorter,
-    /// the value of words sent in MOSI after all `write` has been sent is implementation-defined,
-    /// typically `0x00`, `0xFF`, or configurable.
-    fn transfer(&mut self, read: &mut [W], write: &[W]) -> Result<(), Self::Error>;
-}
-
-impl<T: Transfer<W>, W> Transfer<W> for &mut T {
-    type Error = T::Error;
-
-    fn transfer(&mut self, read: &mut [W], write: &[W]) -> Result<(), Self::Error> {
-        T::transfer(self, read, write)
-    }
-}
-
-/// Blocking transfer with single buffer (in-place)
-pub trait TransferInplace<W = u8> {
+/// Base SPI trait
+///
+/// This just defines the error type, to be used by the other SPI traits.
+pub trait Spi {
     /// Error type
     type Error: crate::spi::Error;
-
-    /// Writes and reads simultaneously. The contents of `words` are
-    /// written to the slave, and the received words are stored into the same
-    /// `words` buffer, overwriting it.
-    fn transfer_inplace(&mut self, words: &mut [W]) -> Result<(), Self::Error>;
 }
 
-impl<T: TransferInplace<W>, W> TransferInplace<W> for &mut T {
+impl<T: Spi> Spi for &mut T {
     type Error = T::Error;
-
-    fn transfer_inplace(&mut self, words: &mut [W]) -> Result<(), Self::Error> {
-        T::transfer_inplace(self, words)
-    }
 }
 
-/// Blocking read
-pub trait Read<W = u8> {
-    /// Error type
-    type Error: crate::spi::Error;
-
+/// Blocking read-only SPI
+pub trait Read<W = u8>: Spi {
     /// Reads `words` from the slave.
     ///
     /// The word value sent on MOSI during reading is implementation-defined,
     /// typically `0x00`, `0xFF`, or configurable.
     fn read(&mut self, words: &mut [W]) -> Result<(), Self::Error>;
+
+    /// Reads all slices in `words` from the slave as part of a single SPI transaction.
+    ///
+    /// The word value sent on MOSI during reading is implementation-defined,
+    /// typically `0x00`, `0xFF`, or configurable.
+    fn read_batch(&mut self, words: &mut [&mut [W]]) -> Result<(), Self::Error> {
+        for buf in words {
+            self.read(buf)?;
```

Comment:
This default impl would not work for a `Read + ManagedCS` impl as `read()` would assert/deassert CS for each slice.